File: typescript/src/pagination.ts
/** Pagination iterators for lab endpoints. */
import type { PageResult } from "./models.js";

export type LabEndpointMeta = {
  id: string;
  method: string;
  style: string;
};

export const LAB_ENDPOINTS: LabEndpointMeta[] = [
  {
    "id": "portal_perks",
    "method": "portalPerks",
    "style": "pageRequest"
  }
];

const CEB_GROUPS = "ABCDEFGHIJKLMNOPQRSTUVWXY".split("");

function asList(raw: unknown): unknown[] {
  if (raw == null) return [];
  if (Array.isArray(raw)) return raw;
  if (typeof raw === "object") {
    const obj = raw as Record<string, unknown>;
    for (const key of ["data", "items", "results", "content", "features", "clusters", "reqTradeSummaryList", "offers", "records", "rows"]) {
      if (Array.isArray(obj[key])) return obj[key] as unknown[];
    }
    return [raw];
  }
  return [raw];
}

export type FetchPage = (...args: any[]) => Promise<unknown>;

export async function* iterPages(
  fetchPage: FetchPage,
  options: {
    style: string;
    start?: number;
    maxPages?: number;
    pageSize?: number;
  },
): AsyncGenerator<PageResult> {
  const style = (options.style || "page_limit").toLowerCase();
  const start = options.start ?? 1;
  const maxPages = options.maxPages ?? 50;
  const pageSize = options.pageSize ?? 50;

  if (["client_slice", "full_download", "client_array", "html_list", "multi_host", "time_window"].includes(style)) {
    const raw = await fetchPage();
    const items = asList(raw);
    if (items.length === 0) {
      yield { page: 0, offset: 0, limit: pageSize, items: [], raw, done: true };
      return;
    }
    for (let i = 0; i < items.length; i += pageSize) {
      const chunk = items.slice(i, i + pageSize);
      const page = Math.floor(i / pageSize);
      yield {
        page,
        offset: i,
        limit: pageSize,
        items: chunk,
        raw: page === 0 ? raw : chunk,
        done: i + pageSize >= items.length,
      };
    }
    return;
  }

  if (style === "group_id") {
    const groups = CEB_GROUPS.slice(0, maxPages);
    for (let idx = 0; idx < groups.length; idx++) {
      const group = groups[idx]!;
      const raw = await fetchPage(group);
      const items = asList(raw);
      yield {
        page: idx,
        offset: idx,
        limit: 1,
        key: group,
        items,
        raw,
        done: idx === groups.length - 1,
      };
    }
    return;
  }

  for (let i = 0; i < maxPages; i++) {
    let page = start + i;
    let offset = (page - 1) * pageSize;
    let key = String(page);
    let raw: unknown;
    if (style === "arcgis") {
      offset = (start > 0 ? start : 0) + i * pageSize;
      page = i;
      key = String(offset);
      raw = await fetchPage(offset, pageSize);
    } else if (style === "pagerequest") {
      const index = (start >= 0 ? start : 0) + i;
      page = index;
      offset = index * pageSize;
      key = String(index);
      raw = await fetchPage(index, pageSize);
    } else {
      raw = await fetchPage(page, pageSize);
    }
    const items = asList(raw);
    const empty = items.length === 0;
    yield { page, offset, limit: pageSize, key, items, raw, done: empty };
    if (empty) break;
    if ((style === "arcgis" || style === "pagerequest") && items.length < pageSize) break;
  }
}

export async function* iterLabEndpoint(
  client: Record<string, FetchPage>,
  endpointId: string,
  options: { start?: number; maxPages?: number; pageSize?: number } = {},
): AsyncGenerator<PageResult> {
  const meta = LAB_ENDPOINTS.find((m) => m.id === endpointId);
  if (!meta) throw new Error(`No lab endpoint ${endpointId} for visa-lk-perks-api-docs`);
  const fn = client[meta.method];
  if (!fn) throw new Error(`Client missing method ${meta.method}`);
  const style = meta.style;
  const fetchPage: FetchPage = async (...args: any[]) => {
    if (style.toLowerCase() === "group_id") return fn(args[0] ?? "A");
    if (["client_slice", "full_download", "client_array", "html_list", "multi_host", "time_window"].includes(style.toLowerCase())) {
      return fn();
    }
    if (args.length >= 2) return fn(args[0], args[1]);
    return fn();
  };
  yield* iterPages(fetchPage, { style, ...options });
}

File: python/visa_lk_perks_api_docs/pagination.py
"""Pagination iterators for lab endpoints.

Generated — regenerate via scripts/scaffold-client-extras.py
"""

from __future__ import annotations

from collections.abc import Callable, Iterator
from typing import Any

from .models import PageResult

LAB_ENDPOINTS: list[dict[str, Any]] = [
  {
    "id": "portal_perks",
    "method": "portal_perks",
    "style": "pageRequest",
    "summary": "VMORC perks; needs siteId + perkTypeRequests body."
  }
]

_CEB_GROUPS = list("ABCDEFGHIJKLMNOPQRSTUVWXY")


def _as_list(raw: Any) -> list[Any]:
    if raw is None:
        return []
    if isinstance(raw, list):
        return raw
    if isinstance(raw, dict):
        for key in (
            "data",
            "items",
            "results",
            "content",
            "features",
            "clusters",
            "reqTradeSummaryList",
            "offers",
            "records",
            "rows",
        ):
            val = raw.get(key)
            if isinstance(val, list):
                return val
        return [raw]
    return [raw]


def _is_empty(raw: Any, items: list[Any]) -> bool:
    if raw is None:
        return True
    if isinstance(raw, str) and not raw.strip():
        return True
    return len(items) == 0


def iter_pages(
    fetch_page: Callable[..., Any],
    *,
    style: str,
    start: int = 1,
    max_pages: int = 50,
    page_size: int = 50,
    extract_items: Callable[[Any], list[Any]] | None = None,
) -> Iterator[PageResult]:
    """Generic pagination iterator.

    ``fetch_page`` signatures by style:
    - page_limit / limit_page: ``(page, limit)``
    - page_number: ``(page_number, size)``
    - arcgis: ``(result_offset, result_record_count)``
    - group_id: ``(group_id,)``  — iterates A–Y
    - pageRequest: ``(index, limit)``  — 0-based index
    - client_slice / full_download / client_array / html_list / multi_host / time_window:
      single fetch then client-side chunking
    """
    extract = extract_items or _as_list
    style_l = (style or "page_limit").lower()

    if style_l in {"client_slice", "full_download", "client_array", "html_list", "multi_host", "time_window"}:
        raw = fetch_page()
        items = extract(raw)
        if not items:
            yield PageResult(page=0, offset=0, limit=page_size, items=[], raw=raw, done=True)
            return
        for i in range(0, len(items), page_size):
            chunk = items[i : i + page_size]
            page = i // page_size
            yield PageResult(
                page=page,
                offset=i,
                limit=page_size,
                items=chunk,
                raw=raw if page == 0 else chunk,
                done=i + page_size >= len(items),
            )
        return

    if style_l == "group_id":
        groups = _CEB_GROUPS[:max_pages]
        for idx, group in enumerate(groups):
            raw = fetch_page(group)
            items = extract(raw)
            yield PageResult(
                page=idx,
                offset=idx,
                limit=1,
                key=group,
                items=items,
                raw=raw,
                done=idx == len(groups) - 1,
            )
        return

    for i in range(max_pages):
        if style_l in {"page_limit", "limit_page"}:
            page = start + i
            raw = fetch_page(page, page_size)
            offset = (page - 1) * page_size
            key = str(page)
        elif style_l == "page_number":
            page = start + i
            raw = fetch_page(page, page_size)
            offset = (page - 1) * page_size
            key = str(page)
        elif style_l == "arcgis":
            offset = (start if start > 0 else 0) + i * page_size
            page = i
            raw = fetch_page(offset, page_size)
            key = str(offset)
        elif style_l == "pagerequest":
            index = (start if start >= 0 else 0) + i
            page = index
            offset = index * page_size
            raw = fetch_page(index, page_size)
            key = str(index)
        else:
            page = start + i
            raw = fetch_page(page, page_size)
            offset = (page - 1) * page_size
            key = str(page)

        items = extract(raw)
        empty = _is_empty(raw, items)
        yield PageResult(
            page=page if style_l != "arcgis" else i,
            offset=offset,
            limit=page_size,
            key=key,
            items=items,
            raw=raw,
            done=empty,
        )
        if empty:
            break
        if style_l == "arcgis" and len(items) < page_size:
            break
        if style_l == "pagerequest" and len(items) < page_size:
            break


def iter_lab_endpoint(
    client: Any,
    endpoint_id: str,
    *,
    start: int = 1,
    max_pages: int = 50,
    page_size: int = 50,
    **call_kwargs: Any,
) -> Iterator[PageResult]:
    """Iterate a catalog lab endpoint by id using the client method."""
    meta = next((m for m in LAB_ENDPOINTS if m["id"] == endpoint_id), None)
    if meta is None:
        raise KeyError(f"No lab endpoint {endpoint_id!r} for visa-lk-perks-api-docs")
    method_name = meta["method"]
    style = meta["style"]
    fn = getattr(client, method_name)

    def fetch_page(*args: Any) -> Any:
        if style.lower() == "group_id":
            return fn(args[0] if args else "A", **call_kwargs)
        if style.lower() in {"client_slice", "full_download", "client_array", "html_list", "multi_host", "time_window"}:
            return fn(**call_kwargs)
        if len(args) == 2:
            return fn(args[0], args[1], **call_kwargs)
        return fn(**call_kwargs)

    return iter_pages(
        fetch_page,
        style=style,
        start=start,
        max_pages=max_pages,
        page_size=page_size,
    )


__all__ = ["LAB_ENDPOINTS", "PageResult", "iter_pages", "iter_lab_endpoint"]
